### ISSUE TITLE
[WinML] [C++/WinRT] Clarify how to share Ort::Env environments with WinRT/WinML instances

### ISSUE THREAD

**Issue by wbudd** [microsoft/onnxruntime#4971]
*This is actually a cross-post of an issue I posted over at [WinML](https://github.com/microsoft/Windows-Machine-Learning/issues/347), but I thought it might be worth asking here too.*

According to [WinML documentation](https://docs.microsoft.com/en-us/windows/ai/windows-ml/#in-box-vs-nuget-winml-solutions), the NuGet WinML solution provides "Direct access to the onnxruntime.dll".

However, all ONNXRuntime functionality requires creation of an `Ort::Env` environment instance as its first order of business, which apparently [can only be created once per process](https://github.com/microsoft/onnxruntime/issues/1430#issuecomment-512968069).

A consequence of that seems to be that if I have WinML inference tasks running in thread **A** through the WinRT API; I am unable to, say, lookup ONNX input/output tensor names/dimension through the ONNXRuntime API in thread **B**—or vice versa— given that I'm neither able to reference WinML's internal `Ort::Env` (*or can I?*), nor can I pass my own `Ort::Env` instance to WinML constructors (*or can I*?). Attempting to use separate instances anyway results in the following error—either thrown by WinML or my side, depending simply on which side creates their instance first:

```Only one instance of LoggingManager created with InstanceType::Default can exist at any point in time.```

Is there anyway to share the same `onnxruntime.dll` with WinRT/WinML, for example by accessing the WinML backend through the ONNXRuntime API instead?

**Comment by pranavsharma** [comment]
There was a bug in OrtEnv creation due to which it was throwing the error you mentioned. It has been [fixed](https://github.com/microsoft/onnxruntime/commit/f0edd074fb6a957b7c40a5135af61654043ed1c2) and will be available as part of the upcoming 1.5 release.

**Comment by wbudd** [comment]
Thanks! Some clarification would be nice though. Does this mean that [the statement](https://github.com/microsoft/onnxruntime/issues/1430#issuecomment-512968069) *"OrtEnv should be created only once, for each process"* is inaccurate (or no longer accurate)? If so, does that mean that within the same process, an internal `Ort::Env` inside the WinML library can now live alongside another `Ort::Env` in an application which also calls Onnx Runtime directly?

**Comment by pranavsharma** [comment]
CreateEnv will always return the same instance of OrtEnv no matter how many times you call it. A bug prevented users from calling CreateEnv more than once; this bug has been fixed. 
Not sure which version of WinML you're using, so I can't comment on that, but if the version that calls CreateEnv is used, you should get the above mentioned behavior. cc @martinb35 to comment on the WinML api.

**Comment by wbudd** [comment]
> cc @martinb35 to comment on the WinML api.

I wouldn't mind a comment here @martinb35 — before the stale bot sweeps yet another issue under the rug.

**Comment by deischi** [comment]
That CreateEnv should be used only once per Session is not very well documented (should I create a separate issue for that?)

From the interface it seems that you can have multiple environments. In most cases this probably does not matter, but if you e.g. would like to use logging functions, or configure a global thread pool you might notice only very late.

Updating the documentation (header file) would be helpful.

**Comment by martinb35** [comment]
@wbudd - you can see the Windows ML call to ORT::CreateEnv [here](https://github.com/microsoft/onnxruntime/blob/897310f6fb709713298efea7ea772687de7fb67a/winml/lib/Api.Ort/OnnxruntimeEnvironment.cpp#L208), which is the same way that @pranavsharma mentioned, so it should return the same instance. Having said that, our primary use cases are either all through the Windows ML interface or all through the OnnxRuntime interface, and we haven't designed/tested (yet) to handle the use case where you use both as your describing. 

@pranavsharma - did you address the comment from @deischi about documenting how to use CreateEnv?

**Comment by wbudd** [comment]
@martinb35 Thanks for confirming that it should be the same instance.

In any case, I came to the conclusion that using the `Microsoft.ML.OnnxRuntime.DirectML` NuGet package instead of the WinML package makes a lot more sense when you also need ONNX Runtime functionality beyond what WinML offers; so that's what I recommend anyone who finds themselves in a similar situation.

**Comment by smk2007** [comment]
@wbudd Yep, it makes more sense to use Microsoft.ML.ONNXRuntime.* packages when "makes a lot more sense when you also need ONNX Runtime functionality beyond what WinML offers"!

However, looking at your original question, it seemed like you wanted to "lookup ONNX input/output tensor names/dimension."

You should be able to look up input/output tensors with the LearningModel.Inputs and Outputs APIs in WinML. Was this not sufficient for you?
